var/spack/repos/builtin/packages/siesta/package.py

```diff
@@ -49,49 +98,209 @@ def flag_handler(self, name, flags):
+
+                        if "mumps" in spec:
+                            libs_arg.append(
+                                "-L{0} -lmumps_common -lzmumps".format(
+                                    self.spec["mumps"].prefix.lib
+                                )
+                            )
+                            fppflags_arg.append("-DSIESTA__MUMPS ")
+
+                        if "+pexsi" in spec:
+                            libs_arg.append(
+                                "-L{0} -lpexsi_linux".format(self.spec["pexsi"].prefix.lib)
+                            )
+                            fppflags_arg.append("-DSIESTA__PEXSI ")
+                            f.write("INCFLAGS += -I{0}/include".format(self.spec["pexsi"].prefix))
+
+                        arch_make.filter("^LIBS =.*", "LIBS = {0}".format(" ".join(libs_arg)))
+                        f.write("FPPFLAGS = {0}".format(" ".join(fppflags_arg)))
 
     def build(self, spec, prefix):
         with working_dir("Obj"):
             make(parallel=False)
-        with working_dir("Obj_trans"):
-            make("transiesta", parallel=False)
-        with working_dir("Util"):
-            sh = which("sh")
-            sh("build_all.sh")
+        if spec.satisfies("@:4.0.2"):
+            with working_dir("Obj_trans"):
+                make("transiesta", parallel=False)
+            with working_dir("Util"):
+                sh = which("sh")
+                sh("build_all.sh")
 
     def install(self, spec, prefix):
         mkdir(prefix.bin)
         with working_dir("Obj"):
             install("siesta", prefix.bin)
-        with working_dir("Obj_trans"):
-            install("transiesta", prefix.bin)
+        if spec.satisfies("@:4.0.2"):
+            with working_dir("Obj_trans"):
+                install("transiesta", prefix.bin)
         for root, _, files in os.walk("Util"):
             for fname in files:
                 fname = join_path(root, fname)
                 if os.access(fname, os.X_OK):
                     install(fname, prefix.bin)
+
+
+class CMakeBuilder(cmake.CMakeBuilder):
+    """Use the new CMake build system to build siesta@5.0.0:."""
+
+    def cmake_args(self):
+        spec = self.spec
+        args = []
+
+        args += ["-DBLAS_LIBRARIES={0}".format(self.spec["blas"].libs.link_flags)]
+        args += ["-DLAPACK_LIBRARIES={0}".format(self.spec["lapack"].libs.link_flags)]
+
+        if "+mpi" in spec:
+            args += [
+                "-DCMAKE_C_COMPILER=%s" % spec["mpi"].mpicc,
+                "-DCMAKE_CXX_COMPILER=%s" % spec["mpi"].mpicxx,
+                "-DCMAKE_Fortran_COMPILER=%s" % spec["mpi"].mpifc,
+            ]
+            args += ["-DSIESTA_WITH_MPI=ON"]
+            args += ["-DSCALAPACK_LIBRARY={0}".format(spec["scalapack"].libs.joined(";"))]
+
+        if "+openmp" in spec:
+            args += ["-DSIESTA_WITH_OPENMP=ON"]
+            if "+cray" in spec:
+                args += ["-DFortran_FLAGS=-fopenmp"]
```

Comment:
As shown above, building the 5.x versions does not work yet due to not setting up the `Obj` directory. As these do not build yet, the support for `buildsystem=cmake` should be moved to a later PR:
```suggestion
```

Reply:
As shown above, you don't need the Obj directory. The sources are different from the older versions.
For more informations : 
https://gitlab.com/siesta-project/siesta/-/blob/dev/ReleaseNotes.md?ref_type=heads#changes

Reply:
Thanks for the explanation. But this means that the way the build_system is selected to be either makefile or `cmake` is in needs of an update to ensure that `makefile` is not used for `@5:`:
```py
build_system(conditional("cmake", when="@5:"), conditional("makefile", when="@:4"))
```

Reply:
````
$ spack spec siesta@4.1.5%gcc@12.1
Input spec
--------------------------------
 -   siesta@4.1.5%gcc@12.1

Concretized
--------------------------------
[+]  siesta@4.1.5%gcc@12.1~cray~debug~elpa~metis+mpi~mumps~netcdf+openmp~pexsi build_system=makefile build_type=Release arch=linux-rhel8-zen4
[e]      ^cray-libsci@23.05.1.4%gcc@12.1~mpi~openmp+shared build_system=generic arch=linux-rhel8-zen4
[e]      ^cray-mpich@8.1.28%gcc@12.1+wrappers build_system=generic arch=linux-rhel8-zen4
[^]      ^gmake@4.4.1%gcc@12.1~guile build_system=generic arch=linux-rhel8-zen4
[+]      ^intel-mkl@2020.4.304%gcc@12.1~ilp64+shared build_system=generic threads=none arch=linux-rhel8-zen4
[+]          ^cpio@2.14%gcc@12.1 build_system=autotools arch=linux-rhel8-zen4
[e]      ^netcdf-c@4.9.0.9%gcc@12.1+blosc~byterange~dap~fsync~hdf4~jna+mpi~nczarr_zip+optimize~parallel-netcdf+pic+shared+szip+zstd build_system=autotools patches=0161eb8 arch=linux-rhel8-zen4
[e]      ^netcdf-fortran@4.9.0.9%gcc@12.1~doc+pic+shared build_system=autotools arch=linux-rhel8-zen4
````
````
spack spec siesta@5.0.1%gcc@12.1
Input spec
--------------------------------
 -   siesta@5.0.1%gcc@12.1

Concretized
--------------------------------
 -   siesta@5.0.1%gcc@12.1~cray~debug~elpa~ipo~metis+mpi~mumps~netcdf+openmp~pexsi build_system=cmake build_type=Release generator=make arch=linux-rhel8-zen4
[^]      ^cmake@3.27.7%gcc@12.1~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-rhel8-zen4
[^]          ^curl@8.4.0%gcc@12.1~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=mbedtls arch=linux-rhel8-zen4
[^]              ^mbedtls@3.3.0%gcc@12.1+pic build_system=makefile build_type=Release libs=static arch=linux-rhel8-zen4
[^]              ^nghttp2@1.57.0%gcc@12.1 build_system=autotools arch=linux-rhel8-zen4
[^]              ^pkgconf@1.9.5%gcc@12.1 build_system=autotools arch=linux-rhel8-zen4
[^]          ^ncurses@6.4%gcc@12.1~symlinks+termlib abi=none build_system=autotools arch=linux-rhel8-zen4
[^]          ^zlib-ng@2.1.4%gcc@12.1+compat+opt build_system=autotools arch=linux-rhel8-zen4
[e]      ^cray-libsci@23.05.1.4%gcc@12.1~mpi~openmp+shared build_system=generic arch=linux-rhel8-zen4
[e]      ^cray-mpich@8.1.28%gcc@12.1+wrappers build_system=generic arch=linux-rhel8-zen4
[^]      ^gmake@4.4.1%gcc@12.1~guile build_system=generic arch=linux-rhel8-zen4
[e]      ^netcdf-c@4.9.0.9%gcc@12.1+blosc~byterange~dap~fsync~hdf4~jna+mpi~nczarr_zip+optimize~parallel-netcdf+pic+shared+szip+zstd build_system=autotools patches=0161eb8 arch=linux-rhel8-zen4
[e]      ^netcdf-fortran@4.9.0.9%gcc@12.1~doc+pic+shared build_system=autotools arch=linux-rhel8-zen4
[+]      ^netlib-scalapack@2.2.0%gcc@12.1~ipo~pic+shared build_system=cmake build_type=Release generator=make patches=072b006,1c9ce5f,244a9aa arch=linux-rhel8-zen4
````
Versions 5.x defaults to '**build_system=cmake**', while versions 4.x uses '**build_system=makefile**'

Reply:
Thanks, but the updated
```py
build_system(conditional("cmake", when="@5.0.0:"), "makefile", default="cmake")
```
still accepts `siesta@5 build_system=makefile` which then fails with `No such file or directory: 'Obj'`.

If you apply this change from https://github.com/CINES-HPC/spack/pull/1, spack would reject that and enforce that only build_system=cmake applies:
```py
build_system(conditional("cmake", when="@5:"), conditional("makefile", when="@:4"))
```
Also, this makes the intention and the facts that there is no choice and the fact that v4 only support `makefile` and v5 only supports `cmake` clear. Your current version lies about that.

As that is better discussed on the other conversation, I resolve this conversation.